File: cpp/ecbenchmark/src/AbstractFunction.cpp
#include "ecbenchmark/AbstractFunction.h"
#include "ecbenchmark/util/CommonMath.h"

#include <sstream>


namespace ecb {

    AbstractFunction::AbstractFunction(const std::string& name,
            const std::string& description, int dimensions,
            scalar minimum, scalar maximum, bool minimization)
    : Function(name), _description(description), _dimensions(dimensions),
    _minimum(minimum), _maximum(maximum), _minimization(minimization) { }

    AbstractFunction::~AbstractFunction() { }

    void AbstractFunction::setDescription(const std::string& description) {
        this->_description = description;
    }

    std::string AbstractFunction::description() const {
        return this->_description;
    }

    void AbstractFunction::setDimensions(int dimensions) {
        this->_dimensions = dimensions;
    }

    int AbstractFunction::dimensions() const {
        return this->_dimensions;
    }

    void AbstractFunction::setMinimumDomain(scalar min) {
        this->_minimum = min;
    }

    scalar AbstractFunction::minimumDomain() const {
        return this->_minimum;
    }

    void AbstractFunction::setMaximumDomain(scalar max) {
        this->_maximum = max;
    }

    scalar AbstractFunction::maximumDomain() const {
        return this->_maximum;
    }

    void AbstractFunction::setDomain(scalar min, scalar max) {
        setMinimumDomain(min);
        setMaximumDomain(max);
    }

    void AbstractFunction::setMinimization(bool min) {
        this->_minimization = min;
    }

    bool AbstractFunction::isMinimization() const {
        return this->_minimization;
    }

    void AbstractFunction::setMaximization(bool max) {
        this->_minimization = !max;
    }

    bool AbstractFunction::isMaximization() const {
        return !this->_minimization;
    }

    //returns positive if a is better than b, negative otherwise

    bool AbstractFunction::isBetter(scalar aValue, scalar bValue) const {
        return isMinimization() ? aValue < bValue : aValue > bValue;
        //        if (isMinimization()) {
        //            return (a - b > 0.0) ? -1 : 1;
        //        }else{
        //            return (a - b < 0.0) ? -1 : 1;
        //        }
    }

    bool AbstractFunction::isBetter(const Solution* a, const Solution* b) const {
        if (a != NULL && b != NULL) {
            return isBetter(a->getObjectiveValue(), b->getObjectiveValue());
        } else {
            throw "NULL pointer exception";
        }
    }

    scalar AbstractFunction::worstValue() const {
        return isMinimization() ? INFINITY : -INFINITY;
    }

    std::string AbstractFunction::toString() const {
        std::stringstream ss;
        ss << name() << "[dimensions= " << dimensions() << "; "
                << "(" << minimumDomain() << ", " << maximumDomain() << "); "
                << "minimization=" << isMinimization() << "]";
        return ss.str();

    }

    std::string AbstractFunction::rString() {
        std::stringstream ss;
        ss << "cols = colorRampPalette(c('black','white'));\n";
        std::vector<scalar> x_y;
        int samples = 100;
        for (int i = 1; i <= samples; ++i) {
            x_y.push_back(CommonMath::Scale(1.0, samples, i, minimumDomain(), maximumDomain()));
        }



        ss << name() + "_x = " + name() + "_y = " + CommonMath::C(x_y) + ";\n";

        std::vector<scalar> z;
        for (size_t i = 0; i < x_y.size(); ++i) {
            for (size_t j = 0; j < x_y.size(); ++j) {
                std::vector<scalar> input;
                input.push_back(x_y[i]);
                input.push_back(x_y[j]);
                z.push_back(f(input));
            }
        }

        ss << name() << "_z = matrix(" << CommonMath::C(z) << ", " << x_y.size() << ", " << x_y.size() << ", "
                << "byrow=TRUE);\n\n";
        ss << "image(" << name() << "_x, " << name() << "_y, " << name() << "_z,"
                << "col=cols(255),axes=FALSE,xlab=NA,ylab=NA,asp=1,main='" << name() << "');\n";

        return ss.str();
    }




}

File: java/ecbenchmark/src/ecbenchmark/AbstractFunction.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package ecbenchmark;

import ecbenchmark.util.CommonMath;
import java.util.ArrayList;
import java.util.List;
import r.util.Datatype;

/**
 *
 * @author jcrada
 */
public abstract class AbstractFunction implements Function {

    protected String name = "", description = "";
    protected int dimensions = 1;
    protected double minimum = Double.NEGATIVE_INFINITY, maximum = Double.POSITIVE_INFINITY;
    protected boolean minimization = true;
    
    public AbstractFunction() {
    }

    public AbstractFunction(String name, String description, int dimensions) {
        this.name = name;
        this.description = description;
        this.dimensions = dimensions;
    }

    public AbstractFunction(String name, String description, int dimensions,
            double minimum, double maximum) {
        this(name, description, dimensions);
        this.minimum = minimum;
        this.maximum = maximum;
        this.minimization = true;
    }

    public AbstractFunction(String name, String description, int dimensions,
            double minimum, double maximum, boolean minimization) {
        this(name, description, dimensions, minimum, maximum);
        this.minimization = minimization;
    }

    

    public void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return this.name;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getDescription() {
        return this.description;
    }

    public void setDimensions(int dimensions) {
        this.dimensions = dimensions;
    }

    public int getDimensions() {
        return this.dimensions;
    }

    public void setMinimum(double min) {
        this.minimum = min;
    }

    public double getMinimum() {
        return this.minimum;
    }

    public void setMaximum(double max) {
        this.maximum = max;
    }

    public double getMaximum() {
        return this.maximum;
    }

    public void setDomain(double min, double max) {
        setMinimum(min);
        setMaximum(max);
    }

    public void setMinimization(boolean min) {
        this.minimization = min;
    }

    public boolean isMinimization() {
        return this.minimization;
    }

    public void setMaximization(boolean max) {
        this.minimization = !max;
    }

    public boolean isMaximization() {
        return !this.minimization;
    }


    public double worstFitness() {
        return isMinimization() ? Double.POSITIVE_INFINITY
                : Double.NEGATIVE_INFINITY;
    }

    boolean isBetter(double fitness_a, double fitness_b) {
        return isMinimization()
                ? fitness_a < fitness_b
                : fitness_a > fitness_b;
    }

    boolean isWorse(double fitness_a, double fitness_b) {
        return isMinimization()
                ? fitness_a > fitness_b
                : fitness_a < fitness_b;
    }

    @Override
    public String toString() {
        return getName() + "[dimensions= " + getDimensions() + "; "
                + "(" + getMinimum() + ", " + getMaximum() + "); "
                + "minimization=" + isMinimization() + "]";
    }

    public String rString() {
        String r = "";
        r += "cols = colorRampPalette(c('black','white'));\n";
        List<Double> x_y = new ArrayList<Double>();
        int samples = 100;
        for (int i = 1; i <= samples; ++i) {
            x_y.add(CommonMath.Scale(1.0, samples, i, getMinimum(), getMaximum()));
        }
        r += getName() + "_x = " + getName() + "_y = " + Datatype.C(x_y) + ";\n";

        List<Double> z = new ArrayList<Double>();
        for (int i = 0; i < x_y.size(); ++i) {
            for (int j = 0; j < x_y.size(); ++j) {
                double[] input = new double[2];
                input[0] = x_y.get(i);
                input[1] = x_y.get(j);
                z.add(f(input));
            }
        }

        r += getName() + "_z = matrix(" + Datatype.C(z) + ", " + x_y.size() + ", " + x_y.size() + ", "
                + "byrow=T);\n\n";
        r += "image(" + getName() + "_x, " + getName() + "_y, " + getName() + "_z,"
                + "col=cols(255),axes=F,xlab=NA,ylab=NA,asp=1,main='" + getName() + "');\n";

        return r;
    }
}
